library/core/src/slice/cmp.rs

```diff
@@ -7,15 +7,20 @@ use crate::mem;
 use super::from_raw_parts;
 use super::memchr;
 
+/// Calls implementation provided memcmp.
+///
+/// Interprets the data as u8.
+///
+/// Returns 0 for equal, < 0 for less than and > 0 for greater
+/// than.
 extern "C" {
-    /// Calls implementation provided memcmp.
-    ///
-    /// Interprets the data as u8.
-    ///
-    /// Returns 0 for equal, < 0 for less than and > 0 for greater
-    /// than.
-    // FIXME(#32610): Return type should be c_int
+    #[cfg(c_int_width = "16")]
+    fn memcmp(s1: *const u8, s2: *const u8, n: usize) -> i16;
+    // Assume we're building the bootstrap compiler on a system with 32-bit c-int width
+    #[cfg(any(bootstrap, c_int_width = "32"))]
     fn memcmp(s1: *const u8, s2: *const u8, n: usize) -> i32;
+    #[cfg(c_int_width = "64")]
+    fn memcmp(s1: *const u8, s2: *const u8, n: usize) -> i64;
```

Comment:
I don't think it's necessary to expand the compiler's public interface for this (introduce `#[cfg(int_width = ...)]`), `type c_int` can be defined just by `cfg`ing on target architectures, like it's done in `library\std\src\os\raw\mod.rs` or in the `libc` crate.

Right now there are only two architectures that differ from the `i32` default - `avr` and `msp430`.